pyface/ui/wx/util/gui_test_assistant.py

```diff
@@ -0,0 +1,272 @@
+#
+#  code based on the traits-enaml.testing.gui_test_assistant
+#
+# ---------------------------------------------------------------------------
+import contextlib
+import threading
+
+import wx
+
+from traits.testing.unittest_tools import UnittestTools
+from traits.testing.unittest_tools import _TraitsChangeCollector as \
+    TraitsChangeCollector
+from pyface.util.guisupport import get_app_wx
+
+from .event_loop_helper import EventLoopHelper, ConditionTimeoutError
+
+
+def find_widget(start, type_, test=None):
+    """Recursively walks the widget tree from widget `start` until it
+       finds a widget of type `type_` (a toolkit widget subclass) that
+        satisfies the provided `test` method.
+
+       Parameters
+       ----------
+       start :
+          The toolkit widget from which to start walking the tree
+       type_ : type
+           A subclass of toolkit widget to use for an initial type filter
+           while walking the tree
+       test : callable
+           A filter function that takes one argument (the current widget being
+           evaluated) and returns either True or False to determine if the
+           widget matches the required criteria.
+
+    """
+    if test is None:
+        test = lambda widget: True
+    if isinstance(start, type_):
+        if test(start):
+            return start
+    for child in start.GetChildren():
+        widget = find_widget(child, type_, test=test)
+        if widget:
+            return widget
+    return None
+
+
+def print_widget_tree(widget, level=0):
+    """ Debugging helper to print out the widget tree starting at a
+    particular `widget`.
+
+    Parameters
+    ----------
+    widget :
+        The root widget in the tree to print.
+    level : int
+        The current level in the tree. Used internally for displaying the
+        tree level.
+    """
+    level = level + 4
+    if level == 0:
+        print
+    print ' '*level, widget
+    for child in widget.GetChildren():
+        print_widget_tree(child, level=level)
+    if level == 0:
+        print
+
+
+class GuiTestAssistant(UnittestTools):
+
+    def setUp(self):
+        # Make sure that we have a running application.
+        self.gui_app = get_app_wx()
+        wx.Log.SetActiveTarget(wx.LogStderr())
+        self.event_loop_helper = EventLoopHelper(gui_app=self.gui_app)
+
+    def tearDown(self):
+        windows = wx.GetTopLevelWindows()
+        titles = []
+        if len(windows) > 0:
+            for window in windows:
+                titles.append(window.GetTitle())
+
+        def _cleanup():
+            for window in wx.GetTopLevelWindows():
+                window.Destroy()
+                wx.WakeUpIdle()
+
+        wx.CallLater(50, _cleanup)
+        self.gui_app.MainLoop()
+        if len(titles) > 0:
+            raise RuntimeError(
+                'Not all windows closed after test method, {}'.format(titles))
+        del self.event_loop_helper
+        del self.gui_app

```

Comment:
Are these about removing references to the GUI objects? Should `del` be called before raising RuntimeError?  I am not sure if these references are kept alive when multiple test cases are being run.

Reply:
yes we should call del before raising